src/testdrive/src/action/consistency.rs

```diff
@@ -0,0 +1,239 @@
+        }
+    }
+}
+
+/// Runs consistency checks against multiple parts of Materialize to make sure we haven't violated
+/// our invariants or leaked resources.
+pub async fn run_consistency_checks(state: &State) -> Result<ControlFlow, anyhow::Error> {
+    let coordinator = check_coordinator(state).await.context("coordinator");
+    let catalog_state = check_catalog_state(state).await.context("catalog state");
+    // TODO(parkmycar): Fix subsources so they don't leak their shards and then add a leaked shards
+    // consistency check.
+
+    // Make sure to report all inconsistencies, not just the first.
+    let mut msg = String::new();
+    if let Err(e) = coordinator {
+        writeln!(&mut msg, "coordinator inconsistency: {e:?}")?;
+    }
+    if let Err(e) = catalog_state {
+        writeln!(&mut msg, "catalog inconsistency: {e:?}")?;
+    }
+
+    if msg.is_empty() {
+        Ok(ControlFlow::Continue)
+    } else {
+        Err(anyhow!("{msg}"))
+    }
+}
+
+/// Checks if a shard in Persist has been tombstoned.
+///
+/// TODO(parkmycar): Run this as part of the consistency checks, instead of as a specific command.
+pub async fn run_check_shard_tombstoned(
+    mut cmd: BuiltinCommand,
+    state: &State,
+) -> Result<ControlFlow, anyhow::Error> {
+    let shard_id = cmd.args.string("shard-id")?;
+    check_shard_tombstoned(state, &shard_id).await?;
+    Ok(ControlFlow::Continue)
+}
+
+/// Asks the Coordinator to run it's own internal consistency checks.
+async fn check_coordinator(state: &State) -> Result<(), anyhow::Error> {
+    let response = Retry::default()
+        .max_duration(Duration::from_secs(2))
+        .retry_async(|_| async {
+            reqwest::get(&format!(
+                "http://{}/api/coordinator/check",
+                state.materialize_internal_http_addr,
+            ))
+            .await
+        })
+        .await
+        .context("querying coordinator")?;
+    if response.status() == StatusCode::NOT_FOUND {
+        tracing::info!("Coordinator consistency check not available");
+        return Ok(());
+    }
+
+    let inconsistencies: serde_json::Value =
+        response.json().await.context("deserialize response")?;
+
+    match inconsistencies {
+        serde_json::Value::String(x) if x.is_empty() => Ok(()),
+        other => Err(anyhow!("coordinator inconsistencies! {other:?}")),
+    }
+}
+
+/// Checks that the in-memory catalog matches what we have persisted on disk.
+async fn check_catalog_state(state: &State) -> Result<(), anyhow::Error> {
+    let maybe_disk_catalog = state
+        .with_catalog_copy(|catalog| catalog.state().clone())
+        .await
+        .map_err(|e| anyhow!("failed to read on-disk catalog state: {e}"))?
+        .map(|catalog| catalog.dump().expect("state must be dumpable"));
+    let Some(disk_catalog) = maybe_disk_catalog else {
+        tracing::warn!("No Catalog state on disk, skipping consistency check");
+        return Ok(());
+    };
```

Comment:
When do we expect this to happen?

Reply:
Same as above, this is never expected so let me upgrade this to an error